src/pybind/mgr/volumes/fs/volume.py

```diff
@@ -0,0 +1,352 @@
+        # create pools
+        r, outs, outb = self.create_pool(metadata_pool, 16, pg_num_min=16, pg_autoscale_factor=4.0)
+        if r != 0:
+            return r, outb, outs
+        r, outb, outs = self.create_pool(data_pool, 8)
+        if r != 0:
+            return r, outb, outs
+        # create filesystem
+        r, outb, outs = self.create_filesystem(volname, metadata_pool, data_pool)
+        if r != 0:
+            log.error("Filesystem creation error: {0} {1} {2}".format(r, outb, outs))
+            return r, outb, outs
+        # create mds
+        return self.create_mds(volname)
+
+    def delete_volume(self, volname):
+        """
+        delete the given module (tear down mds, remove filesystem)
+        """
+        # Tear down MDS daemons
+        try:
+            completion = self.mgr.remove_stateless_service("mds", volname)
+            self.mgr._orchestrator_wait([completion])
+            orchestrator.raise_if_exception(completion)
+        except (ImportError, orchestrator.OrchestratorError):
+            log.warning("OrchestratorError, not tearing down MDS daemons")
+        except Exception as e:
+            # Don't let detailed orchestrator exceptions (python backtraces)
+            # bubble out to the user
+            log.exception("Failed to tear down MDS daemons")
+            return -errno.EINVAL, "", str(e)
+
+        # In case orchestrator didn't tear down MDS daemons cleanly, or
+        # there was no orchestrator, we force the daemons down.
+        if self.volume_exists(volname):
+            r, outb, outs = self.set_mds_down(volname)
+            if r != 0:
+                return r, outb, outs
+            r, outb, outs = self.remove_filesystem(volname)
+            if r != 0:
+                return r, outb, outs
+        else:
+            log.warning("Filesystem already gone for volume '{0}'".format(volname))
+        metadata_pool, data_pool = self.gen_pool_names(volname)
+        r, outb, outs = self.remove_pool(metadata_pool)
+        if r != 0:
+            return r, outb, outs
+        return self.remove_pool(data_pool)
+
+    def list_volumes(self):
+        result = []
+        fs_map = self.mgr.get("fs_map")
+        for f in fs_map['filesystems']:
+            result.append({'name': f['mdsmap']['fs_name']})
+        return 0, json.dumps(result, indent=2), ""
+
+    ### subvolume operations
+
+    def create_subvolume(self, volname, subvolname, groupname, size):
+        ret = 0, "", ""
+        try:
+            if not self.volume_exists(volname):
+                raise VolumeException(
+                    -errno.ENOENT, "Volume '{0}' not found, create it with `ceph fs " \
+                    "volume create` before trying to create subvolumes".format(volname))
+            with SubVolume(self.mgr, fs_name=volname) as sv:
+                spec = SubvolumeSpec(subvolname, groupname)
+                if not sv.get_group_path(spec, create_default=True):
```

Comment:
I don't follow, why you need 'create_default', here and elsewhere? It creates the default subvolume group? The create_subvolume() will create the default subvolume group.

Reply:
That *might* have been necessary for supporting subvolume/group list commands. Since those are not in this PR now, I need to recheck if this is still required. WIll get to this when I address comments.